File: typescript/Linked Lists/linked_list_reversal.ts
import { ListNode } from './ds';

/* Definition of ListNode:
class ListNode {
    val: number; next: ListNode | null;
    constructor(val: number){
        this.val = val;
        this.next = null;
    }
}*/

function linkedListReversal(head: ListNode | null): ListNode | null {
    let currNode: ListNode | null = head;
    let prevNode: ListNode | null = null;
    // Reverse the direction of each node's pointer until 'currNode' 
    // is null.
    while (currNode !== null){
        const nextNode: ListNode | null = currNode.next;
        currNode.next = prevNode;
        prevNode = currNode;
        currNode = nextNode;
    }
    // 'prevNode' will be pointing at the head of the reversed linked 
    // list.
    return prevNode;
}

File: csharp/Linked Lists/LinkedListReversal.cs
using ds.ListNode;

/*
	Definition of ListNode:
	class ListNode
	{
		public ListNode(int val, ListNode next = null)
		{
			Val = val;
			Next = next;
		}

		public int Val { get; set; }

		public ListNode Next { get; set; }
	}
*/

public class Solution
{
	public ListNode LinkedListReversal(ListNode head)
	{
		ListNode currentNode = head;
		ListNode prevNode = null;

		// Reverse the direction of each node's pointer until 'currentNode' is null.
		while (currentNode != null)
		{
			ListNode nextNode = currentNode.Next;
			currentNode.Next = prevNode;
			prevNode = currentNode;
			currentNode = nextNode;
		}

		// 'prevNode' will be pointing at the head of the reversed linked list.
		return prevNode;
	}
}
